// @storybook/core-server/dist/esm/utils/stories-json.js
import "core-js/modules/es.promise.js";
import fs from 'fs-extra';
import EventEmitter from 'events';
import { normalizeStories } from '@storybook/core-common';
import debounce from 'lodash/debounce';
import { StoryIndexGenerator } from './StoryIndexGenerator';
import { watchStorySpecifiers } from './watch-story-specifiers';
import { useEventsAsSSE } from './use-events-as-sse';
var INVALIDATE = 'INVALIDATE';
export var DEBOUNCE = 50;
export async function extractStoriesJson(outputFile, normalizedStories, options) {
  var generator = new StoryIndexGenerator(normalizedStories, options);
  await generator.initialize();
  var index = await generator.getIndex();
  await fs.writeJson(outputFile, index);
}
export async function useStoriesJson(router, options, workingDir = process.cwd()) {
  var normalizedStories = normalizeStories(await options.presets.apply('stories'), {
    configDir: options.configDir,
    workingDir: workingDir
  });
  var features = await options.presets.apply('features');
  var generator = new StoryIndexGenerator(normalizedStories, {
    configDir: options.configDir,
    workingDir: workingDir,
    storiesV2Compatibility: !(features !== null && features !== void 0 && features.breakingChangesV7) && !(features !== null && features !== void 0 && features.storyStoreV7)
  }); // Wait until someone actually requests `stories.json` before we start generating/watching.
  // This is mainly for testing purposes.

  var started = false;
  var invalidationEmitter = new EventEmitter();
  var maybeInvalidate = debounce(function () {
    return invalidationEmitter.emit(INVALIDATE);
  }, DEBOUNCE, {
    leading: true
  });

  async function ensureStarted() {
    if (started) return;
    started = true;
    watchStorySpecifiers(normalizedStories, {
      workingDir: workingDir
    }, function (specifier, path, removed) {
      generator.invalidate(specifier, path, removed);
      maybeInvalidate();
    });
    await generator.initialize();
  }

  var eventsAsSSE = useEventsAsSSE(invalidationEmitter, [INVALIDATE]);
  router.use('/stories.json', async function (req, res) {
    await ensureStarted();
    if (eventsAsSSE(req, res)) return;

    try {
      var index = await generator.getIndex();
      res.header('Content-Type', 'application/json');
      res.send(JSON.stringify(index));
    } catch (err) {
      res.status(500);
      res.send(err.message);
    }
  });
}
