// @storybook/store/dist/modern/StoryStore.js
import "core-js/modules/es.array.reduce.js";

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import memoize from 'memoizerific';
import mapValues from 'lodash/mapValues';
import pick from 'lodash/pick';
import global from 'global';
import { StoryIndexStore } from './StoryIndexStore';
import { ArgsStore } from './ArgsStore';
import { GlobalsStore } from './GlobalsStore';
import { processCSFFile } from './processCSFFile';
import { prepareStory } from './prepareStory';
import { HooksContext } from './hooks';
import { normalizeInputTypes } from './normalizeInputTypes';
import { inferArgTypes } from './inferArgTypes';
import { inferControls } from './inferControls';
const {
  FEATURES
} = global;
// TODO -- what are reasonable values for these?
const CSF_CACHE_SIZE = 1000;
const STORY_CACHE_SIZE = 10000;

function normalizeProjectAnnotations(_ref) {
  let {
    argTypes,
    globalTypes,
    argTypesEnhancers
  } = _ref,
      annotations = _objectWithoutPropertiesLoose(_ref, ["argTypes", "globalTypes", "argTypesEnhancers"]);

  return Object.assign({}, argTypes && {
    argTypes: normalizeInputTypes(argTypes)
  }, globalTypes && {
    globalTypes: normalizeInputTypes(globalTypes)
  }, {
    argTypesEnhancers: [...(argTypesEnhancers || []), inferArgTypes, // inferControls technically should only run if the user is using the controls addon,
    // and so should be added by a preset there. However, as it seems some code relies on controls
    // annotations (in particular the angular implementation's `cleanArgsDecorator`), for backwards
    // compatibility reasons, we will leave this in the store until 7.0
    inferControls]
  }, annotations);
}

export class StoryStore {
  constructor() {
    this.storyIndex = void 0;
    this.importFn = void 0;
    this.projectAnnotations = void 0;
    this.globals = void 0;
    this.args = void 0;
    this.hooks = void 0;
    this.cachedCSFFiles = void 0;
    this.processCSFFileWithCache = void 0;
    this.prepareStoryWithCache = void 0;
    this.initializationPromise = void 0;
    this.resolveInitializationPromise = void 0;

    this.getStoriesJsonData = () => {
      const value = this.getSetStoriesPayload();
      const allowedParameters = ['fileName', 'docsOnly', 'framework', '__id', '__isArgsStory'];
      const stories = mapValues(value.stories, story => Object.assign({}, pick(story, ['id', 'name', 'title']), {
        importPath: this.storyIndex.stories[story.id].importPath
      }, !(FEATURES !== null && FEATURES !== void 0 && FEATURES.breakingChangesV7) && {
        kind: story.title,
        story: story.name,
        parameters: Object.assign({}, pick(story.parameters, allowedParameters), {
          fileName: this.storyIndex.stories[story.id].importPath
        })
      }));
      return {
        v: 3,
        stories
      };
    };

    this.globals = new GlobalsStore();
    this.args = new ArgsStore();
    this.hooks = {}; // We use a cache for these two functions for two reasons:
    //  1. For performance
    //  2. To ensure that when the same story is prepared with the same inputs you get the same output

    this.processCSFFileWithCache = memoize(CSF_CACHE_SIZE)(processCSFFile);
    this.prepareStoryWithCache = memoize(STORY_CACHE_SIZE)(prepareStory); // We cannot call `loadStory()` until we've been initialized properly. But we can wait for it.

    this.initializationPromise = new Promise(resolve => {
      this.resolveInitializationPromise = resolve;
    });
  }

  setProjectAnnotations(projectAnnotations) {
    // By changing `this.projectAnnotations, we implicitly invalidate the `prepareStoryWithCache`
    this.projectAnnotations = normalizeProjectAnnotations(projectAnnotations);
    const {
      globals,
      globalTypes
    } = projectAnnotations;
    this.globals.set({
      globals,
      globalTypes
    });
  }

  initialize({
    storyIndex,
    importFn,
    cache = false
  }) {
    this.storyIndex = new StoryIndexStore(storyIndex);
    this.importFn = importFn; // We don't need the cache to be loaded to call `loadStory`, we just need the index ready

    this.resolveInitializationPromise();
    if (cache) this.cacheAllCSFFiles(true);
  } // This means that one of the CSF files has changed.
  // If the `importFn` has changed, we will invalidate both caches.
  // If the `storyIndex` data has changed, we may or may not invalidate the caches, depending
  // on whether we've loaded the relevant files yet.


  async onStoriesChanged({
    importFn,
    storyIndex
  }) {
    if (importFn) this.importFn = importFn;
    if (storyIndex) this.storyIndex.stories = storyIndex.stories;

    if (this.cachedCSFFiles) {
      await this.cacheAllCSFFiles(false);
    }
  } // To load a single CSF file to service a story we need to look up the importPath in the index


  loadCSFFileByStoryId(storyId, {
    sync = false
  } = {}) {
    const {
      importPath,
      title
    } = this.storyIndex.storyIdToEntry(storyId);
    const moduleExportsOrPromise = this.importFn(importPath);

    if (sync) {
      // NOTE: This isn't a totally reliable way to check if an object is a promise
      // (doesn't work in Angular), but it's just to help users in an error case.
      if (Promise.resolve(moduleExportsOrPromise) === moduleExportsOrPromise) {
        throw new Error(`importFn() returned a promise, did you pass an async version then call initialize({sync: true})?`);
      } // We pass the title in here as it may have been generated by autoTitle on the server.


      return this.processCSFFileWithCache(moduleExportsOrPromise, title);
    }

    return Promise.resolve(moduleExportsOrPromise).then(moduleExports => // We pass the title in here as it may have been generated by autoTitle on the server.
    this.processCSFFileWithCache(moduleExports, title));
  }

  loadAllCSFFiles(sync) {
    const importPaths = {};
    Object.entries(this.storyIndex.stories).forEach(([storyId, {
      importPath
    }]) => {
      importPaths[importPath] = storyId;
    });
    const csfFileList = Object.entries(importPaths).map(([importPath, storyId]) => ({
      importPath,
      csfFileOrPromise: sync ? this.loadCSFFileByStoryId(storyId, {
        sync: true
      }) : this.loadCSFFileByStoryId(storyId, {
        sync: false
      })
    }));

    function toObject(list) {
      return list.reduce((acc, {
        importPath,
        csfFile
      }) => {
        acc[importPath] = csfFile;
        return acc;
      }, {});
    }

    if (sync) {
      return toObject(csfFileList.map(({
        importPath,
        csfFileOrPromise
      }) => ({
        importPath,
        csfFile: csfFileOrPromise
      })));
    }

    return Promise.all(csfFileList.map(async ({
      importPath,
      csfFileOrPromise
    }) => ({
      importPath,
      csfFile: await csfFileOrPromise
    }))).then(toObject);
  }

  cacheAllCSFFiles(sync) {
    if (sync) {
      this.cachedCSFFiles = this.loadAllCSFFiles(true);
      return null;
    }

    return this.loadAllCSFFiles(false).then(csfFiles => {
      this.cachedCSFFiles = csfFiles;
    });
  } // Load the CSF file for a story and prepare the story from it and the project annotations.


  async loadStory({
    storyId
  }) {
    await this.initializationPromise;
    const csfFile = await this.loadCSFFileByStoryId(storyId, {
      sync: false
    });
    return this.storyFromCSFFile({
      storyId,
      csfFile
    });
  } // This function is synchronous for convenience -- often times if you have a CSF file already
  // it is easier not to have to await `loadStory`.


  storyFromCSFFile({
    storyId,
    csfFile
  }) {
    const storyAnnotations = csfFile.stories[storyId];

    if (!storyAnnotations) {
      throw new Error(`Didn't find '${storyId}' in CSF file, this is unexpected`);
    }

    const componentAnnotations = csfFile.meta;
    const story = this.prepareStoryWithCache(storyAnnotations, componentAnnotations, this.projectAnnotations);
    this.args.setInitial(story.id, story.initialArgs);
    this.hooks[story.id] = this.hooks[story.id] || new HooksContext();
    return story;
  } // If we have a CSF file we can get all the stories from it synchronously


  componentStoriesFromCSFFile({
    csfFile
  }) {
    return Object.keys(csfFile.stories).map(storyId => this.storyFromCSFFile({
      storyId,
      csfFile
    }));
  } // A prepared story does not include args, globals or hooks. These are stored in the story store
  // and updated separtely to the (immutable) story.


  getStoryContext(story) {
    return Object.assign({}, story, {
      args: this.args.get(story.id),
      globals: this.globals.get(),
      hooks: this.hooks[story.id]
    });
  }

  cleanupStory(story) {
    this.hooks[story.id].clean();
  }

  extract(options = {
    includeDocsOnly: false
  }) {
    if (!this.cachedCSFFiles) {
      throw new Error('Cannot call extract() unless you call cacheAllCSFFiles() first.');
    }

    return Object.entries(this.storyIndex.stories).reduce((acc, [storyId, {
      importPath
    }]) => {
      const csfFile = this.cachedCSFFiles[importPath];
      const story = this.storyFromCSFFile({
        storyId,
        csfFile
      });

      if (!options.includeDocsOnly && story.parameters.docsOnly) {
        return acc;
      }

      acc[storyId] = Object.entries(story).reduce((storyAcc, [key, value]) => {
        if (typeof value === 'function') {
          return storyAcc;
        }

        if (Array.isArray(value)) {
          return Object.assign(storyAcc, {
            [key]: value.slice().sort()
          });
        }

        return Object.assign(storyAcc, {
          [key]: value
        });
      }, {
        args: story.initialArgs
      });
      return acc;
    }, {});
  }

  getSetStoriesPayload() {
    const stories = this.extract({
      includeDocsOnly: true
    });
    const kindParameters = Object.values(stories).reduce((acc, {
      title
    }) => {
      acc[title] = {};
      return acc;
    }, {});
    return {
      v: 2,
      globals: this.globals.get(),
      globalParameters: {},
      kindParameters,
      stories
    };
  }

  raw() {
    return Object.values(this.extract()).map(({
      id
    }) => this.fromId(id));
  }

  fromId(storyId) {
    if (!this.cachedCSFFiles) {
      throw new Error('Cannot call fromId/raw() unless you call cacheAllCSFFiles() first.');
    }

    let importPath;

    try {
      ({
        importPath
      } = this.storyIndex.storyIdToEntry(storyId));
    } catch (err) {
      return null;
    }

    const csfFile = this.cachedCSFFiles[importPath];
    const story = this.storyFromCSFFile({
      storyId,
      csfFile
    });
    return Object.assign({}, story, {
      storyFn: update => {
        const context = Object.assign({}, this.getStoryContext(story), {
          viewMode: 'story'
        });
        return story.unboundStoryFn(Object.assign({}, context, update));
      }
    });
  }

}
