// @storybook/preview-web/dist/modern/WebView.js
import global from 'global';
import { logger } from '@storybook/client-logger';
import AnsiToHtml from 'ansi-to-html';
import dedent from 'ts-dedent';
const {
  document
} = global;
const layoutClassMap = {
  centered: 'sb-main-centered',
  fullscreen: 'sb-main-fullscreen',
  padded: 'sb-main-padded'
};
const classes = {
  MAIN: 'sb-show-main',
  NOPREVIEW: 'sb-show-nopreview',
  ERROR: 'sb-show-errordisplay'
};
const ansiConverter = new AnsiToHtml({
  escapeXML: true
});
export class WebView {
  constructor() {
    this.currentLayoutClass = void 0;
  }

  // Get ready to render a story, returning the element to render to
  prepareForStory(story) {
    this.showStory();
    this.applyLayout(story.parameters.layout);
    document.documentElement.scrollTop = 0;
    document.documentElement.scrollLeft = 0;
    return this.storyRoot();
  }

  storyRoot() {
    return document.getElementById('root');
  }

  prepareForDocs() {
    this.showMain();
    this.showDocs();
    this.applyLayout('fullscreen');
    return this.docsRoot();
  }

  docsRoot() {
    return document.getElementById('docs-root');
  }

  applyLayout(layout = 'padded') {
    if (layout === 'none') {
      document.body.classList.remove(this.currentLayoutClass);
      this.currentLayoutClass = null;
      return;
    }

    this.checkIfLayoutExists(layout);
    const layoutClass = layoutClassMap[layout];
    document.body.classList.remove(this.currentLayoutClass);
    document.body.classList.add(layoutClass);
    this.currentLayoutClass = layoutClass;
  }

  checkIfLayoutExists(layout) {
    if (!layoutClassMap[layout]) {
      logger.warn(dedent`The desired layout: ${layout} is not a valid option.
         The possible options are: ${Object.keys(layoutClassMap).join(', ')}, none.`);
    }
  }

  showErrorDisplay({
    message = '',
    stack = ''
  }) {
    document.getElementById('error-message').innerHTML = ansiConverter.toHtml(message);
    document.getElementById('error-stack').innerHTML = ansiConverter.toHtml(stack);
    document.body.classList.remove(classes.MAIN);
    document.body.classList.remove(classes.NOPREVIEW);
    document.body.classList.add(classes.ERROR);
  }

  showNoPreview() {
    var _this$storyRoot, _this$docsRoot;

    document.body.classList.remove(classes.MAIN);
    document.body.classList.remove(classes.ERROR);
    document.body.classList.add(classes.NOPREVIEW); // In storyshots this can get called and these two can be null

    (_this$storyRoot = this.storyRoot()) === null || _this$storyRoot === void 0 ? void 0 : _this$storyRoot.setAttribute('hidden', 'true');
    (_this$docsRoot = this.docsRoot()) === null || _this$docsRoot === void 0 ? void 0 : _this$docsRoot.setAttribute('hidden', 'true');
  }

  showMain() {
    document.body.classList.remove(classes.NOPREVIEW);
    document.body.classList.remove(classes.ERROR);
    document.body.classList.add(classes.MAIN);
  }

  showDocs() {
    this.storyRoot().setAttribute('hidden', 'true');
    this.docsRoot().removeAttribute('hidden');
  }

  showStory() {
    this.docsRoot().setAttribute('hidden', 'true');
    this.storyRoot().removeAttribute('hidden');
  }

}
